include/awkward/Identity.h

```diff
@@ -1,89 +1,103 @@
-    virtual const std::shared_ptr<Identity> to64() const = 0;
-    virtual const std::string tostring_part(const std::string indent, const std::string pre, const std::string post) const = 0;
-    virtual const std::shared_ptr<Identity> getitem_range_nowrap(int64_t start, int64_t stop) const = 0;
-    virtual const std::shared_ptr<Identity> shallow_copy() const = 0;
-    virtual const std::shared_ptr<Identity> getitem_carry_64(const Index64& carry) const = 0;
-    virtual const std::shared_ptr<Identity> withfieldloc(const FieldLoc& fieldloc) const = 0;
-    virtual int64_t value(int64_t row, int64_t col) const = 0;
-
-    const std::string tostring() const;
-
-  protected:
-    const Ref ref_;
-    const FieldLoc fieldloc_;
-    int64_t offset_;
-    int64_t width_;
-    int64_t length_;
-  };
-
-  template <typename T>
-  class IdentityOf: public Identity {
-  public:
-    IdentityOf<T>(const Ref ref, const FieldLoc fieldloc, int64_t width, int64_t length)
-        : Identity(ref, fieldloc, 0, width, length)
-        , ptr_(std::shared_ptr<T>(length*width == 0 ? nullptr : new T[(size_t)(length*width)], awkward::util::array_deleter<T>())) { }
-    IdentityOf<T>(const Ref ref, const FieldLoc fieldloc, int64_t offset, int64_t width, int64_t length, const std::shared_ptr<T> ptr)
-        : Identity(ref, fieldloc, offset, width, length)
-        , ptr_(ptr) { }
-
-    const std::shared_ptr<T> ptr() const { return ptr_; }
-
-    virtual const std::string classname() const;
-    virtual const std::string location_at(int64_t at) const;
-    virtual const std::shared_ptr<Identity> to64() const;
-    virtual const std::string tostring_part(const std::string indent, const std::string pre, const std::string post) const;
-    virtual const std::shared_ptr<Identity> getitem_range_nowrap(int64_t start, int64_t stop) const;
-    virtual const std::shared_ptr<Identity> shallow_copy() const;
-    virtual const std::shared_ptr<Identity> getitem_carry_64(const Index64& carry) const;
-    virtual const std::shared_ptr<Identity> withfieldloc(const FieldLoc& fieldloc) const;
-    virtual int64_t value(int64_t row, int64_t col) const;
-
-    const std::vector<T> getitem_at(int64_t at) const;
-    const std::vector<T> getitem_at_nowrap(int64_t at) const;
-    const std::shared_ptr<Identity> getitem_range(int64_t start, int64_t stop) const;
-
-  private:
-    const std::shared_ptr<T> ptr_;
-  };
-
-  typedef IdentityOf<int32_t> Identity32;
-  typedef IdentityOf<int64_t> Identity64;
-}
+class Identity {
+public:
+  typedef int64_t Ref;
+  typedef std::vector<std::pair<int64_t, std::string>> FieldLoc;
+
+  static Ref newref();
+  static std::shared_ptr<Identity> none() {
+    return std::shared_ptr<Identity>(nullptr);
+  }
+
+  Identity(const Ref ref, const FieldLoc fieldloc, int64_t offset,
+           int64_t width, int64_t length)
+      : ref_(ref), fieldloc_(fieldloc), offset_(offset), width_(width),
+        length_(length) {}
```

Comment:
But not this. I made all those parameters line up for a reason. Some of these classes have long parameter lists (on purpose) and formatting them as a vertical line makes them easier to read.